src/Compilers/VisualBasic/Portable/CommandLine/VisualBasicCompiler.vb

```diff
@@ -143,11 +142,21 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                  Arguments.CompilationOptions.
                      WithMetadataReferenceResolver(referenceDirectiveResolver).
                      WithAssemblyIdentityComparer(assemblyIdentityComparer).
-                     WithStrongNameProvider(strongNameProvider).
                      WithXmlReferenceResolver(xmlFileResolver).
                      WithSourceReferenceResolver(sourceFileResolver))
 
-            Return result
+            Dim loggingOperations = New LoggingStrongNameFileSystem(touchedFilesLogger)
+            Dim searchPaths = Arguments.KeyFileSearchPaths
+
+            Dim fallback =
+                (result.Feature("BypassStrongName") <> Nothing) Or
+                (result.Options.CryptoKeyContainer <> Nothing)
+
+            Return result.WithOptions(
+                result.Options.WithStrongNameProvider(
+                    If(fallback,
+                        CType(New DesktopStrongNameProvider(searchPaths, Nothing, loggingOperations), StrongNameProvider),
+                        CType(New PortableStrongNameProvider(searchPaths, loggingOperations), StrongNameProvider))))
```

Comment:
This entire block looks identical to the corresponding C# code. Consider extracting a helper method.

Reply:
We should be able to extract a helper method to calculate the `StrongNameProvider`, for use here and in `CSharpCompiler`. Also, we're allocating a second `Compilation` instance above when using `result.WithOptions()`. Consider reverting to a single `VisualBasicCompilation.Create` call. Similarly for `CSharpCompilation.Create`.
```
// Helper method in CommandLineArguments
internal StrongNameProvider GetStrongNameProvider(StrongNameFileSystem fileSystem)
{
    bool fallback = ParseOptionsCore.Features.ContainsKey("UseLegacyStrongNameProvider") ||
        CompilationOptionsCore.CryptoKeyContainer != null;
    return fallback ?
        (StrongNameProvider)new DesktopStrongNameProvider(KeyFileSearchPaths, null, fileSystem) :
        (StrongNameProvider)new PortableStrongNameProvider(KeyFileSearchPaths, fileSystem);
}
```
```
Dim strongNameProvider = Arguments.GetStrongNameProvider(New LoggingStrongNameFileSystem(touchedFilesLogger))
Dim result = VisualBasicCompilation.Create(
    Arguments.CompilationName,
    trees,
    resolvedReferences,
    Arguments.CompilationOptions.
        ...
        WithStrongNameProvider(strongNameProvider).
        ...)
```

```